salt/cloud/clouds/azurearm.py

```diff
@@ -301,10 +312,12 @@ def get_conn(client_type):
             "secret", get_configured_provider(), __opts__, search_global=False
         )
         conn_kwargs.update({"client_id": client_id, "secret": secret, "tenant": tenant})
-    else:
-        username = config.get_cloud_config_value(
-            "username", get_configured_provider(), __opts__, search_global=False
-        )
+
+    username = config.get_cloud_config_value(
+        "username", get_configured_provider(), __opts__, search_global=False
+    )
+
+    if username:
```

Comment:
This one should be much more straightforward - there should be a test that returns none for username and enforces that neither username or password was passed to get_client. And then a test that ensures that if the username was returned that whatever username/password values were provided *were* passed to get_client.